src/primitives/hrp.rs

```diff
@@ -0,0 +1,420 @@
+    #[cfg(feature = "alloc")]
+    pub fn to_lowercase(&self) -> String { self.lowercase_char_iter().collect() }
+
+    /// Creates a byte iterator over the ASCII byte values of this HRP.
+    ///
+    /// If an uppercase HRP was parsed during object construction then this iterator will yield
+    /// uppercase ASCII `char`s. For lowercase bytes see [`Self::lowercase_byte_iter`]
+    pub fn byte_iter(&self) -> ByteIter {
+        ByteIter { iter: self.buf.iter(), size: self.size, n: 0 }
+    }
+
+    /// Creates a character iterator over the ASCII characters of this HRP.
+    ///
+    /// If an uppercase HRP was parsed during object construction then this iterator will yield
+    /// uppercase ASCII `char`s. For lowercase bytes see [`Self::lowercase_char_iter`].
+    pub fn char_iter(&self) -> CharIter { CharIter { iter: self.byte_iter() } }
+
+    /// Creates a lowercase iterator over the bytes (ASCII characters) of this HRP.
+    pub fn lowercase_byte_iter(&self) -> LowercaseByteIter {
+        LowercaseByteIter { iter: self.byte_iter() }
+    }
+
+    /// Creates a lowercase iterator over the bytes (ASCII characters) of this HRP.
+    pub fn lowercase_char_iter(&self) -> LowercaseCharIter {
+        LowercaseCharIter { iter: self.lowercase_byte_iter() }
+    }
+
+    /// Returns the length (number of characters) of the human-readable part.
+    ///
+    /// Guaranteed to be between 1 and 83 inclusive.
+    pub fn len(&self) -> usize { self.size }
+
+    /// The human-readable part is guaranteed to be between 1-83 characters.
+    pub fn is_empty(&self) -> bool { false }
+}
+
+/// Displays the human-readable part.
+///
+/// If an uppercase HRP was parsed during object construction then the returned string will be
+/// in uppercase also. For a lowercase string see [`Self::to_lowercase`].
+impl fmt::Display for Hrp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for c in self.char_iter() {
+            f.write_char(c)?;
+        }
+        Ok(())
+    }
+}
+
+/// Case insensitive comparison.
+impl Ord for Hrp {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.lowercase_byte_iter().cmp(other.lowercase_byte_iter())
+    }
+}
+
+/// Case insensitive comparison.
+impl PartialOrd for Hrp {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> { Some(self.cmp(other)) }
+}
+
+/// Case insensitive comparison.
+impl PartialEq for Hrp {
+    fn eq(&self, other: &Self) -> bool {
+        self.lowercase_byte_iter().eq(other.lowercase_byte_iter())
+    }
+}
+
+impl Eq for Hrp {}
+
+impl core::hash::Hash for Hrp {
+    fn hash<H: core::hash::Hasher>(&self, h: &mut H) { self.buf.hash(h) }
+}
+
+/// Iterator over bytes (ASCII values) of the human-readable part.
+///
+/// ASCII byte values as they were initially parsed (i.e., in the original case).
+pub struct ByteIter<'b> {
+    iter: Iter<'b, u8>,
+    size: usize, // Total size of the iterator.
+    n: usize,    // Number of characters yielded so far.
```

Comment:
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

Why do we need to keep track of this stuff? Why can't `slice::Iter` do it for us?

Reply:
Ha, overthought that didn't I.

Reply:
Damn, I should trust myself more when reading reviews, we keep track of this because if we use `slice::Iter` it includes all the zeros in the hrp internal buffer.

Reply:
Then you should create the `slice::Iter` from a slice of the correct size :)

Reply:
oooo, I should trust reviewers more ... :)